xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/resources/xwiki-focusedplaceholder/plugin.js

```diff
@@ -0,0 +1,268 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+(function() {
+    'use strict';
+
+    var ATTRIBUTE_NAME = 'data-xwiki-focusedplaceholder';
+
+    CKEDITOR.plugins.add('xwiki-focusedplaceholder', {
+
+        /**
+         * Styles that would be applied to the editor by the placeholder text when visible.
+         *
+         * @property {String}
+         */
+        placeholderStyle: '[' + ATTRIBUTE_NAME + ']::before {' +
+            'position: absolute;' +
+            'opacity: .8;' +
+            'color: #aaa;' +
+            'content: attr( ' + ATTRIBUTE_NAME + ' );' +
+            '}' +
+            '.cke_wysiwyg_div[' + ATTRIBUTE_NAME + ']::before {' +
+            'margin-top: 1em;' +
+            '}',
+
+        onLoad: function() {
+            // Placeholder style
+            CKEDITOR.addCss(this.placeholderStyle);
+        },
+
+        beforeInit: function(editor) {
+            // Default plugin configuration, to be overriden by other plugins
+            editor.config.focusedplaceholder = {
+                // Default placeholder for content sectioning, text and table HTML tags
+                placeholder: {
+                    _fallback: "",
+                    body: "xwiki-focusedplaceholder.placeholder.body", // Body
```

Comment:
I don't think the user can generate new `<body>` tags either, however the user can move the caret to a position where it is only within a body tag and not inside a p tag. If we remove placeholders for the body tags, no placeholder will appear when the user erases the whole document for instance.

In the same way, I think that the other tags may appear in a document even if not directly created by CKEditor. The XWiki Syntax provides means to insert html code in the document, a user or extension might introduce the html tags you are suggesting to remove in the document.
Is there a significant overhead if we configure them in case they might be used?

Reply:
> I don't think the user can generate new <body> tags either, however the user can move the caret to a position where it is only within a body tag and not inside a p tag.

![20903-body](https://github.com/xwiki/xwiki-platform/assets/28761965/0c5c3d08-1442-4538-88d8-a496a323f676)
Even if the user gets out of the \<p> , the \<body> is way up here and not something you get close to when editing.
What you actually edit with CKEditor AFAIK is not the layout, but the page content and its title. So everything around (menus, header, footer, side panels, ...) is outside of the editing scope. For some tag names it's debatable if we can include it in the user content (aside especially), but for the few I mentioned above, if it's possible to use them in the editor, it's a major ticket we need to raise. 

> In the same way, I think that the other tags may appear in a document even if not directly created by CKEditor. The XWiki Syntax provides means to insert html code in the document, a user or extension might introduce the html tags you are suggesting to remove in the document.

If they can insert a \<body>, I'd consider it a bug. \<body> and the tag names mentioned above are a bit special, as they should only be used once in a HTML document. If we can use them in user generated content, do we have safeguards to avoid using them multiple times, and what if we already use them outside of the interface?
I don't think they are necessary to write down some page content, so I'm pretty sure CKEditor doesn't give an easy way to use them.

> Is there a significant overhead if we configure them in case they might be used?

Each tag element would need a translation for each language. It's not that much but it adds up quickly for translation contributors :)
Moreover, I think it's unclear to not make a distinction between the tag names that can get generated and those which can't. This might generate more work down the line. If I stumbled across this a few months ago, I might have thought that all tag names could be generated by CKEditor and searched more to see if I should report the potential issue mentioned above.
I think a comment explaining why you're putting every tag name (or not) would help.

Reply:
> Even if the user gets out of the `<p>` , the `<body>` is way up here and not something you get close to when editing.

![body-placeholder](https://github.com/xwiki/xwiki-platform/assets/132468278/e49401df-c108-499c-b32a-01a6a641170b)
The `<body>` is not that far away when using the iframe editor (Which an user can access in advanced mode by clicking the WYSIWYG Editor button from the dropdown edit menu. Apparently the placeholder currently updates before the `<p>` tag is inserted, I will try to find a better event to update the placeholder on to take the `<p>` insertion into account.

Thank you, that also lead me to notice that I was adding event listeners on the DOM Document, which is not what we want for the inline editor, I will find the correct element to add the listeners to in the ckeditor documentation.

> Each tag element would need a translation for each language. It's not that much but it adds up quickly for translation contributors :)

That is something I did not think about, indeed. :sweat_smile:
I can fix the update for late automatic `<p>` insertions, remove the `body` tag from the configured elements
and add a comment explaining that not all the tags can be created by ckeditor but that the caret might be moved into them. What do you think? Should we discuss it with other committers on the XWiki matrix channel?

Reply:
> A somewhat relevant reference: https://ckeditor.com/docs/ckeditor5/latest/features/html/general-html-support.html#enabling-all-html-features

@Sereza7 we're using CKEditor 4 not 5.

> Even if the user gets out of the <p> , the <body> is way up here and not something you get close to when editing.

In-place editing is not the only usage of CKEditor in XWiki. You're forgetting the **standalone** WYSIWYG edit mode (currently used when creating a new wiki page) and the Text Area object properties that are editable with the CKEditor. In these cases the **editable** element is the body element, because the edited content is loaded in an iframe.

> If they can insert a <body>, I'd consider it a bug. <body> and the tag names mentioned above are a bit special, as they should only be used once in a HTML document.

The placeholders @douakli is adding are not about content generation. First you need to understand that WYSIWYG edit mode and CKEditor are separate things. The WYSIWYG edit mode is **using** the CKEditor, but CKEditor is not limited to the WYSIWYG edit mode. CKEditor is a generic rich text editing widget and so it needs to support many uses cases. Nothing prevents me from activating the CKEditor in-line (non-iframe based) on a footer, header or main element. Even if we don't have this use case right now there's no reason to limit this just because the WYSIWYG edit mode doesn't need it. Again, the CKEditor Integration module provides an API for extension developers to use CKEditor in XWiki, outside of the WYSIWYG edit mode.

> Each tag element would need a translation for each language. It's not that much but it adds up quickly for translation contributors :)

I don't think the number of translation messages is a problem.

Reply:
Thanks for the explanation and the demo! I did not understand our use of CKEditor enough.
It's much clearer now. :slightly_smiling_face: